### DISCUSSION TITLE
Should we remove `__webpack_require__.el`?

### DISCUSSION THREAD

**Discussion by ahabhgk** [web-infra-dev/rspack#4699] [Core]
Should we remove `__webpack_require__.el` and alignment `block_promise` with webpack?

motivation:
1. `__webpack_require__.el` is inconsistent with webpack
2. already caused more inconsistent, and we don't how many inconsistent it will cause in the future
    - AsyncDependenciesBlockIdentifier.from
3. lead bigger output, for example:

    ```js
    // entry: a.js
    import("./shared.js")
    ```
    ```js
    // entry: b.js
    import("./shared.js")
    ```
    the `__webpack_require__.el` map of both a.js and b.js will be:
    ```js
    __webpack_require__.el = { "./a.js@0:21": ["shared"], "./b.js@0:21": ["shared"] }
    ```
    which contain duplicate key

    this is caused by `__webpack_require__.el` is actually generated at runtime module codegen, don't  know which block actually belong to current runtime, so we can only iterate all blocks, and get its chunkGroup.runtime, if the chunkGroup.runtime is the superset of the current runtime, then we include it into the `__webpack_require__.el` map

    for example, the chunkGroup of block `"./a.js@0:21"` and block `"./b.js@0:21"` are pointed to same chunkGroup `"shared"`, and its runtime is `["a", "b"]`, so it is included into the `__webpack_require__.el` map of a.js and b.js

    
4. the reason of why we create `__webpack_require__.el` is for generate the correct output assets in incremental build, but actually, incremental build only affect the creation of module graph, the incorrect output assets is caused by the cache of codegen didn't invalidate correctly

**Comment by hardfist** [comment]
if we could solve the incremental build problem, then strong +1 remove the `__webpack_require__.el`, since it's actually a public api of module, it will be very difficult when it's relied by other runtime implementation

**Reply by ahabhgk** [reply]
I think we can solve it, will have a try once I finish MF related stuff
